### DISCUSSION TITLE
Using `Rodauth::Rails.rodauth` as an "API"

### DISCUSSION THREAD

**Discussion by bjeanes** [janko/rodauth-rails#44] [General]
Hey @janko,

This is a continuation of the discussion that started [on Reddit](https://www.reddit.com/r/ruby/comments/jx5wre/adding_authentication_in_rails_6_with_rodauth/gcyh784/) last November, which I believe [prompted you](https://github.com/janko/rodauth-rails/issues/15#issuecomment-731848517) to add `Rodauth::Rails.rodauth`.

In that thread, I linked to a gist which [demonstrated](https://gist.github.com/bjeanes/ead1e6b0e7d729e7db06d2650a849362#file-authentication-rb-L6-L26) what I was doing to `new` up a Roduath auth class instance as an API.

Today, I _finally_ have tried to switch to using your upstream method and immediately hit a few obstacles. I was going to open a GH issue initially (which I still may do) but also wanted to re-visit the larger discussion.

First up, the issues I hit with trying to use `Rodauth::Rails.rodauth`:

* Anything which calls `session` raises:
   ```sh-session
   > Rodauth::Rails.rodauth(:user).session
   Roda::RodaError: You're missing a session handler, try using the sessions plugin.
   from /bundle/gems/roda-3.44.0/lib/roda/request.rb:375:in `session'
   ```
* Anything which calls `param` or `request.params` raises: `RuntimeError: Missing rack.input`
  ``` sh-session
   > Rodauth::Rails.rodauth(:user).param("login")
   RuntimeError: Missing rack.input
   from /bundle/gems/rack-2.2.3/lib/rack/request.rb:442:in `POST'
   ```

Obviously, we are not in a real request setting and I understand that setting a session value (non-cookie session stores aside) won't have any end effect for a user. However, given how much of Roda internals are side-effecting, it isn't practical to be able to call any useful internal methods without touching things like the session.

Furthermore, unless one creates a _separate_ slimmer config for API usage (and consequently risks bugs due to inconsistencies with the primary config), inevitably there will be `before_*`/`after_*` hooks which may read params or read&write session store.

Lastly, Rodauth core features vary with regard to using `account_id` and [using `session_value`](https://github.com/jeremyevans/rodauth/search?q=%22session_value%22) (sometimes for good reason and other times perhaps for historical reasons--I'm not sure). This poses a very real need to address the session store situation and perhaps even pre-filling its contents. 

In the Reddit thread, you had picked up on my `account_from_id` auth method, which merely sets `@account = account_ds(id).first`, but I've found I also need to have `session[session_key] = id` set, so that features using `session_value` are functional. Another example I've encountered is with [`active_sessions` feature](https://rodauth.jeremyevans.net/rdoc/files/doc/active_sessions_rdoc.html) where the [`active_sessions_ds` is defined in terms of `session_value`](https://github.com/jeremyevans/rodauth/blob/3b661c19160f826187d7034520eb3171b57f192a/lib/rodauth/features/active_sessions.rb#L169-L172), causing me grief when wanting to delete a user's active sessions after an admin changes the email address. This comes up regularly with compromised accounts (for which we allow users to prove their identity to regain access).

Anyway, I was able to work around the params issue by _mutating_ the internal Rack env, like so:

```ruby
rodauth = Rodauth::Rails.rodauth(configuration_name)
rodauth.scope.env.merge!({
  Rack::QUERY_STRING => "",
  Rack::RACK_REQUEST_QUERY_STRING => "",
  Rack::RACK_REQUEST_QUERY_HASH => {},
  Rack::RACK_REQUEST_FORM_HASH => {},
})
```

I think setting the query string and form hash out of the box makes sense.

NOTE: in my usage, I actually am passing non-empty params and encoding to query string with `URI.encode_www_form`, because I have hooks which do things like read `param(login_param)` during audit log metadata (i.e. to capture old and new logins). Without having a separate config, I need to contend with this during "API" usage too. I've run into this need often enough, that there is a case for improving `Rodauth::Rails.rodauth` to accept params out of the box, but we can defer that discussion.

For the issue with touching session, I don't understand it and hitting that after fixing the params issue is what led me to leave this post. It gets really into the weeds with my understanding of the separation between Roda and Rodauth.

----

Now, as for the broader discussion of "programmable Rodauth"...  

There are certainly things which `rodauth-rails` could do to make this use-case better, but I think there are some real challenges in the current spelling of Rodauth internals that would require a larger and serious discussion involving Jeremy.

Chiefly is the issue of the atomicity around `before_x`, `x`, `after_x`. Rodauth's out of the box auditing functionality relies on `after_*` being called, for one, and custom config in either `before_*` or `after_*` might affect the behaviour of the unit as a whole (perhaps by `throw`ing, `raise`ing or altering state). Clearly, if an admin changes a users email this can't be done in the request's Rodauth instance or it will have side effects for the admin's session, so it must be done in a quarantined and virtual instance a la `Rodauth::Rails.rodauth` and I don't think anybody would argue that this shouldn't also leave an audit trail.

However, we wouldn't want to simply duplicate the `before_x`, `x`, `after_x` triplets inside other API because sometimes they are not so simple, such as [`change_login` being fallable and `after_change_login` not running in that condition](https://github.com/jeremyevans/rodauth/blob/3b661c19160f826187d7034520eb3171b57f192a/lib/rodauth/features/change_login.rb#L46-L51).

Ultimately, I think we can make Rodauth nice to use for these out-of-request use cases without Jeremy's support to de-couple the core stateful changes from the request life-cycle as a whole. By stateful here, I don't mean things like `@account` so much as things like inserting a password hash into the DB. Ideally, all features would be re-worked to be a discrete API and the Roda-exposed feature routes would use the same API as would a use case outside of that request. That's big work and I'm not sure Jeremy would even be interested in a move in that direction.

Personally, this is the primary (and almost only) grievance I have with Rodauth as a whole. It has the best-designed data model of anything in Rubyland but hasn't really accounted for the use cases that would exist to _support_ the human beings who are the users in that data model.

**Comment by janko** [comment]
Hi @bjeanes, thanks for continuing this discussion 👍🏻 

Based on what you said, it seems that making methods like `#param` and `#session` work would solve most of the problems. It's reassuring that setting certain Rack env values solved the params problem, then the developer can handle empty params in hooks and other rodauth settings. The session shouldn't be difficult to solve, I think setting `rack.session` env key to an empty hash would do the trick (though it's probably a bit more involved than that).

As for the account ID in the session, I'm thinking about an `:account` keyword argument, which would accept an Active Record model instance, and set both `@account` and `session[session_key]` on the returned Rodauth instance.

So, for current account, params, and session, the `Rodauth::Rails.rodauth` method could be extended with the following:

```rb
Rodauth::Rails.rodauth # account, params, and session are empty
Rodauth::Rails.rodauth(params: { foo: "bar" }) # set mock params
Rodauth::Rails.rodauth(account: Account.find(account_id)) # set @account and session[session_key]
Rodauth::Rails.rodauth(session: { two_factor_auth_setup: true }) # set mock session
Rodauth::Rails.rodauth(env: { ... }) # add more values to Rack env hash
```

-----

Regarding not having to call `before_*` and `after_*` hooks manually when calling Rodauth operations, provided that hooks are named the same as the operation (which [isn't always the case](https://github.com/jeremyevans/rodauth/blob/1a42415968d299a9bfb54707bd71d8258c737067/lib/rodauth/features/verify_login_change.rb#L161-L166)), one could have a wrapper method like `#call`, where `call(:something)` would call `before_something`, `something`, and `after_something`. I remember exploring that a few months ago, and working on a few improvements in the Rodauth internals to support that better (mainly extracting some operations), I would like to pick it up again. Is that something that would improve the situation?

I also wanted to mention that over the past several weeks I've been working on a [mixin for the account AR model](https://github.com/janko/rodauth-rails/compare/model), which defines a `password` attribute & validations (that work regardless of whether the password hash is stored in a column or a separate table), and associations for feature tables. This would simplify inserting the password hash (`account.password = "secret"`) or deleting the account's active sessions (`account.active_session_keys.delete_all`). I'm still not 100% sure it's the right path, mainly whether it won't open a slippery slope regarding how much of Rodauth functionality should be supported through the Active Record model, but hopefully it won't.

**Reply by janko** [reply]
I've implemented the following `Rodauth::Rails.rodauth` options on the [`instance-helpers`](https://github.com/janko/rodauth-rails/compare/instance-helpers) branch:

* `:query` for query params and `:form` for form params (in case the distinction needs to be made)
* `:session` for session values
* `:account` which sets `@account` and `session[:account_id]`
* `:env` for setting additional env values

The request parameters and session now default to empty hash instead of raising an exception. Could you try it out and let me know if it simplifies things for you?

**Reply by bjeanes** [reply]
I really like the proposed API surface you've done here. It's not far off from a helper I have:

``` ruby
  def self.rodauth_for(user_or_admin, params: {}, &block)
    config_name =
      case user_or_admin
      when User
        :user
      when Admin
        :admin
      else
        raise "expected User or Admin"
      end

    params = {'login' => user_or_admin.email}.merge(params)

    rodauth(config_name, params: params) do
      account_from_id(user_or_admin.id)
      session[session_key] = user_or_admin.id
      return instance_eval(&block) if block_given?
      self
    end
```

...where `rodauth` here is my equivalent of your `Rodauth::Rails.rodauth`, more-or-less. The block reminds me of another aspect here, which is that there are some methods which are `private` in Rodauth, so the API I ended up with always takes a block yielded inside the Rodauth auth instance.

This represents probably more friction with Rodauth's design internally, but I'll come back to this case in a moment.

---

> I also wanted to mention that over the past several weeks I've been working on a [mixin for the account AR model](https://github.com/janko/rodauth-rails/compare/model), which defines a `password` attribute & validations

This is very interesting. ~I had no idea about `Rodauth::Rails.model` until now.~ (I see, you just added it 🤦🏻‍♂️) I have my own `Authentication::Account` model but it is annotated as `readonly`. I am curious to switch to use this to have convenient access to the full model hierarchy. _However_, at a glance, it would seam `include Rodauth::Rails.model(:user)` mightn't work for me as it's a singular model for both configs (`:user` and `:admin`) for me. I'll have to experiment...

> I'm still not 100% sure it's the right path, mainly whether it won't open a slippery slope regarding how much of Rodauth functionality should be supported through the Active Record model, but hopefully it won't.

Yeah, I am a tad concerned about this. `account.active_session_keys.delete_all` is precisely the case I _want_ to avoid, because it would not trigger hooks or cause Rodauth auditing. It's spooky-action-at-a-distance against the data model of Rodauth, from Rodauth's perspective. It certainly would be more familiar to Rails developers but I know in my case, where I have become the resident Rodauth feature expert and maintainer, that other devs would simply do everything through the AR surface area, because it is familiar. I fear that would cause unexpected inconsistencies down the line.

> Regarding not having to call `before_*` and `after_*` hooks manually when calling Rodauth operations, provided that hooks are named the same as the operation (which [isn't always the case](https://github.com/jeremyevans/rodauth/blob/1a42415968d299a9bfb54707bd71d8258c737067/lib/rodauth/features/verify_login_change.rb#L161-L166)), one could have a wrapper method like `#call`, where `call(:something)` would call `before_something`, `something`, and `after_something`. I remember exploring that a few months ago, and working on a few improvements in the Rodauth internals to support that better (mainly extracting some operations), I would like to pick it up again. Is that something that would improve the situation?

That's interesting... I don't know that it would generically work though. In addition to the case you've cited here where the naming scheme is not mirrored, it (and a few others, such as [`reset_password`](https://github.com/jeremyevans/rodauth/blob/1a42415968d299a9bfb54707bd71d8258c737067/lib/rodauth/features/reset_password.rb#L145-L150)) cause another issue: when _other_ methods are called or state is set in between these method triplets.

Furthermore, both those issues are joined by a third, in the case of the [`otp` feature](https://github.com/jeremyevans/rodauth/blob/1a42415968d299a9bfb54707bd71d8258c737067/lib/rodauth/features/otp.rb#L118-L124); the `before_x` and `x` are run _conditionally_, there is no `after_x` apparently, but there is a _differently_ named after hook which runs when the condition fails (`after_otp_authentication_failure`). This feature is not likely one which would be exhausted via external programming, but its presence is still worth examining when thinking about solutions.

So, while certainly better than nothing, it does re-enforce my belief that we'd need Jeremy on-side to really make this smooth.

Thank you for the responses and for the changes you've made so far. I'll test them in a few hours (I just woke up, here in Australia).

**Reply by bjeanes** [reply]
> The request parameters and session now default to empty hash instead of raising an exception. Could you try it out and let me know if it simplifies things for you?

@janko this really tidies things up immensely for me. ~So far, so good. But I haven't run the whole test suite yet, just the likely offenders.~ All passed.

My complicated mess evaporates into a much more reasonable:

```rb
module Authentication
  # Rodauth is pretty coupled to being in a request context, but this will provide the 
  # minimum necessary to be able to generate correct URLs and perform basic administrative 
  # actions on user accounts
  def self.rodauth(configuration_name = :user, **rest, &block)
    rodauth = Rodauth::Rails.rodauth(configuration_name, **rest)
    return rodauth.instance_eval(&block) if block_given?
    rodauth
  end

  def self.rodauth_for(user_or_admin, form: {}, **rest, &block)
    config_name =
      case user_or_admin
      when User
        :user
      when Admin
        :admin
      else
        raise "expected User or Admin"
      end

    rodauth(
      config_name,
      account: Authentication::Account.find(user_or_admin.id),
      form: {'login' => user_or_admin.email}.merge(form),
      &block
    )
  end

  # ...
end
```

Basically, this allows me two things not covered by rodauth-rails now:

1. Taking a block to avoid `.send()` (try writing a method which _confirms_ an email change programmatically -- which I only use in tests, but still...)
2. Take an application `User` or `Admin` object and choose the correct Rodauth config and pre-configure `@account` with the corresponding account record.

:tada:

**Reply by bjeanes** [reply]
Oh mate, I think taking `env` might just solve one of my big outstanding issues actually.

The service I migrated from Devise to Rodauth back in November is one with a pretty vulnerable user group who get attacked, doxxed, etc., pretty regularly. As such, we lean on the auditing subsystem of Rodauth _heavily_.

However, we also want to audit fully when admins act on behalf of a user so that if an admin makes an error or admin access is gained by an attacker, we have the history. BUT, because this was not in a real request, my auditing of user-agent, IP address, etc., were all absent.

This experiment will have to wait until I am next working in a few days, but I _think_ I might able to resolve the issue by passing `request.env` from Rails' into `Rodauth::Rails.rodauth` which should fill in the missing context quite nicely.

I'm quite excited about this actually...

**Reply by bjeanes** [reply]
Yup, passing `request.env` worked an absolute treat!

**Reply by bjeanes** [reply]
Another example of inconsistent triplets of `before_x`/`x`/`after_x` is removing 2FA:

```rb
before_two_factor_disable # "disable"
two_factor_remove         # "remove"
after_two_factor_disable  # "disable
```

**Reply by janko** [reply]
I've released 0.13.0 with the enhanced `Rodauth::Rails.rodauth`. I chose not to add the automatic configuration detection from the AR model class, because I didn't feel like it would be so common, and also multiple configurations can theoretically use the same accounts table.

Regarding calling Rodauth actions along with hooks, I began experimenting with extracting this code into methods that can be more easily callable directly on the Rodauth instance. You can see the idea [here](https://github.com/janko/rodauth/commit/63b6637d03f545ab9fc7e1cf39bec2fc2a1151f1). This wouldn't rely on hooks following the same naming convention. Also, it seems to me like something Jeremy might accept, because we're extracting code that acts as a single unit into its own methods, which is always beneficial for general design. What do you think about that approach?

**Reply by bjeanes** [reply]
> I've released 0.13.0 with the enhanced `Rodauth::Rails.rodauth`

🙌🏻 🙏🏻 Excellent news.

> I chose not to add the automatic configuration detection from the AR model class, because I didn't feel like it would be so common, and also multiple configurations can theoretically use the same accounts table.

Yeah, agreed. At the very least, it should wait until the need for it is clearer and stronger, if at all. I'm certainly in that camp with multiple configs using same underlying tables so it was hard to see how, if at all, I could use it.

> Regarding calling Rodauth actions along with hooks, I began experimenting with extracting this code into methods that can be more easily callable directly on the Rodauth instance. You can see the idea [here](https://github.com/janko/rodauth/commit/63b6637d03f545ab9fc7e1cf39bec2fc2a1151f1). This wouldn't rely on hooks following the same naming convention. Also, it seems to me like something Jeremy might accept, because we're extracting code that acts as a single unit into its own methods, which is always beneficial for general design. What do you think about that approach?

Oh nice! This is pretty much what I was imagining. I hope you're right that Jeremy would be on-board because I think it would make it a lot cleaner and more powerful without sacrificing any simplicity upstream. In fact, I think it increases simplicity!

It also strengthens something for me. Because I want to record the `actor` and a `reason` in audit logs for these actions-done-by-another-account, I am hooking into `hook_action` for all `:before` hooks (as a heuristic for "mutative action") to enforce the presence of a known `actor`. But it allowed for an error where external code didn't call the `before_*` action. Making these atomic units should make that more resilient.

Let me know if I can help move it a long in anyway, whether testing, contributing, or participating in discussion with Jeremy. I was contemplating a PR along these lines, but wanted to have this conversation first to get my head back in the right context and because you'd have a better sense as to whether Jeremy would be receptive.

Thanks again for everything!

**Reply by janko** [reply]
I've just opened a [thread](https://groups.google.com/g/rodauth/c/Ugod2tnbUXQ) on the rodauth google group to discuss this, so we can go from there 🙂

**Reply by janko** [reply]
I've ran across some issues when trying to support controller callbacks running around internal requests. The problem is, in order to assign rodauth instances in rack env, I would need to assign instances of the *internal subclasses* of rodauth classes. Otherwise the regular rodauth instances complain that they are missing things from `env`, such as the session handler.

I gave it some thought, and I actually don't think controller callbacks should run at all during internal requests. An internal request is not an actual request, it's a programmatic call to the Rodauth API, it just happens to be implemented as a rack request internally. Not running callbacks could cause issues with view rendering, as templates might rely e.g. on instance variables set in the callbacks, but we should be able to avoid it by skipping view rendering altogether during internal requests (the rendered HTML is not used anyway AFAICT).

How does that sound? Do you see any issues in skipping callbacks during internal requests?

**Reply by bjeanes** [reply]
That's amazing--I had the same _exact_ same thought while tossing around in bed restlessly last night.

I entirely agree. If rodauth-rails is able to just step aside for these internal requests, I think that's entirely workable. Huge +1

**Reply by janko** [reply]
Great, so here is the PR that adds support for internal requests – https://github.com/janko/rodauth-rails/pull/53. With these changes I was able to successfully send internal requests in the demo app. Please test it out and let me know if anything is missing.

**Reply by bjeanes** [reply]
Will do! I just bought a house yesterday so I can't promise it will get my attention before next week, but I'll try to take a close look ASAP.
